crates/shared/src/order_quoting.rs

```diff
@@ -530,6 +548,78 @@ impl OrderQuoting for OrderQuoter {
     }
 }
 
+/// Handles whether or not quotes need to be verified.
+enum QuoteVerification {
+    /// Quotes get trusted blindly.
+    Never,
+    /// Enforce quotes only when the user has the required sell token balance.
+    /// If a user has the required sell token balance and a quote can still not
+    /// be verified the trade likely involves a weird token that solvers are not
+    /// able to handle correctly.
+    WhenFunded(Arc<dyn BalanceFetching>),
+}
+
+impl QuoteVerification {
+    /// Makes sure a quote was verified according to the configured rule.
+    async fn check(
+        &self,
+        estimate: &Estimate,
+        parameters: &QuoteParameters,
+        sell_amount: U256,
+    ) -> Result<(), CalculateQuoteError> {
+        let Some(verification) = &parameters.verification else {
+            return Ok(()); // no verification requested
+        };
+        if estimate.verified {
+            return Ok(()); // successfully verified
+        }
+        let balance_fetcher = match &self {
+            QuoteVerification::Never => return Ok(()),
+            QuoteVerification::WhenFunded(fetcher) => fetcher.as_ref(),
+        };
+
+        let balance =
+            match Self::get_balance(balance_fetcher, verification, parameters.sell_token).await {
+                Ok(balance) => balance,
+                Err(err) => {
+                    tracing::warn!(?err, "could not fetch balance for verification");
+                    return Err(CalculateQuoteError::QuoteNotVerified);
+                }
+            };
+
+        if balance >= sell_amount {
+            // Quote could not be verified although user has the required balance.
+            // This likely indicates a weird token that solvers are not able to handle.
+            return Err(CalculateQuoteError::QuoteNotVerified);
+        }
+
+        Ok(())
+    }
+
+    async fn get_balance(
+        balance_fetcher: &dyn BalanceFetching,
+        verification: &Verification,
+        token: H160,
+    ) -> Result<U256> {
+        let query = Query {
+            owner: verification.from,
+            token,
+            source: verification.sell_token_source,
+            interactions: verification
+                .pre_interactions
+                .iter()
+                .map(|i| InteractionData {
+                    target: i.target,
+                    value: i.value,
+                    call_data: i.data.clone(),
```

Comment:
This clone is expensive and could be avoided by passing by value the `parameters.verification` in the `check()` function. The caller of the `check()` function does not need `verification` afterwards, so [here](https://github.com/cowprotocol/services/pull/2621/files#diff-3f8b6079ce50ae3ab0fe449d91c7e68922cae4023902961c892b58f91ddc49b3R436) `parameters` could be destroy in order to move `verification` into `check()` and then destroying it in this line where the clone would not be necessary.

Reply:
Since `calculate_quote()` already takes `&QuoteParameters` by reference the caller would also have to be changed to avoid this clone. It's possible and I tried it out but overall I think it leads to worse code since we'd have to store a few struct fields in member variables because we need those after the code loses the ownership of that struct.
Also the vast majority of quotes does not have pre-interactions anyway and when they do they are fairly small. I would not worry about cloning a few bytes here and there when optimizing that away would lead to code that is trickier to read.